### PR TITLE
[FlagGems Operator Development Competition] Add svd operator

### PR DESCRIPTION
**Pull request by lcthej**
### PR Category
  [Operator] [OP Test] [Benchmark]

  ### Type of Change
  [New Feature] [Performance Optimization]

  ### Description
  Add a CUDA/Triton implementation for `torch.svd`.

  This PR includes:
  - `svd` operator registration
  - Triton implementations for rank-1, rank-2, small Jacobi, 4x4, Gram/eigh, and 16-rank paths
  - fallback to ATen for unsupported modes/dtypes/shapes
  - unit tests for accuracy, fallback modes, non-contiguous input, empty input, and complex fallback
  - core benchmark coverage for SVD shapes
  - Triton autotuner compatibility cleanup to avoid deprecated warmup/rep/use_cuda_graph warnings

  ### Issue
  N/A

  ### Progress
  - [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
  - [ ] Change is responded to an issue.
  - [x] Change is fully covered by a UT.

  ### Performance
  Warning-clean local validation:

  - `ruff check src/flag_gems/utils/libentry.py src/flag_gems/ops/svd.py tests/test_svd.py benchmark/test_svd.py`: passed
  - `python3 -m py_compile src/flag_gems/utils/libentry.py src/flag_gems/ops/svd.py tests/test_svd.py benchmark/test_svd.py`: passed
  - `pytest -q tests/test_svd.py -m svd -W error`: 14 passed
  - `pytest -q -s benchmark/test_svd.py --level core --warmup 5 --iter 10 --record log -m svd -W error`: 1 passed

  Core benchmark result: 33/33 SVD benchmark rows succeeded. Latest warning-clean local run showed most shapes faster than `torch.svd`, with speedups ranging from near parity to over `7000x` depending on shape.

### CODE REVIEW

File: src/flag_gems/utils/libentry.py [hunk 1]
```diff
@@ -281,6 +281,39 @@ def __init__(
...
+                arg_names,
+                configs,
+                key,
+                reset_to_zero,
+                restore_value,
+                pre_hook=pre_hook,
+                post_hook=post_hook,
+                prune_configs_by=prune_configs_by,
+                do_bench=do_bench,
+            )
```
Comment (bin913) [inline src/flag_gems/utils/libentry.py:319]: What is the reason for modifying this code section?

**Comment by bin913** [comment]
@lcthej  Please fix code-style and resolve conflicts

File: src/flag_gems/utils/libentry.py (hunk 1 above)
Comment (lcthej) [reply src/flag_gems/utils/libentry.py:319]: This change is for Triton autotuner benchmark compatibility, not for SVD math itself.

On newer Triton versions, forwarding `warmup`, `rep`, and `use_cuda_graph` directly through the autotuner path can trigger deprecated benchmark-argument warnings. The update keeps the old behavior for older Triton versions, and for Triton 3.2+ converts those values into the `do_bench` callback expected by the newer autotuner API.

This was needed so the SVD tests and benchmark can run warning-clean under `-W error`. The change is intended to be backward-compatible: older Triton paths still receive the legacy autotuner arguments, while Triton 3.2+ receives an explicit benchmark function.

**Comment by lcthej** [comment]
Adding a coverage/performance checklist for the RaceDetail test-completeness items, limited to the CUDA/NVIDIA evidence available for this PR.

### Coverage checklist

| Dimension | Covered in this PR |
| --- | --- |
| API surface | `torch.svd(self, some=True, compute_uv=True)` registration |
| Shape scale | Core benchmark includes 33 SVD rows covering tiny batched matrices, tall/skinny, wide/short, square, large square, and batched square cases |
| Implementation paths | Rank-1, rank-2, small Jacobi, 4x4, Gram/eigh, and rank-16 paths |
| Fallback modes | ATen fallback for unsupported modes, dtypes, complex inputs, and unsupported shapes |
| Correctness / edge cases | Accuracy tests, fallback modes, non-contiguous input, empty input, and complex fallback |
| Benchmark coverage | Adds `benchmark/test_svd.py` and `benchmark/core_shapes.yaml` entry |
| Validation commands | `ruff check src/flag_gems/utils/libentry.py src/flag_gems/ops/svd.py tests/test_svd.py benchmark/test_svd.py`; `python3 -m py_compile src/flag_gems/utils/libentry.py src/flag_gems/ops/svd.py tests/test_svd.py benchmark/test_svd.py`; `pytest -q tests/test_svd.py -m svd -W error`; `pytest -q -s benchmark/test_svd.py --level core --warmup 5 --iter 10 --record log -m svd -W error` |
| Validation result | ruff passed; py_compile passed; unit tests `14 passed`; core benchmark `1 passed`; 33/33 benchmark rows succeeded |

### Performance evidence

Measured on the SVD core benchmark:

| evidence item | result |
| --- | --- |
| command | `pytest -q -s benchmark/test_svd.py --level core --warmup 5 --iter 10 --record log -m svd -W error` |
| benchmark rows | 33/33 SVD rows succeeded |
| aggregate observation | most shapes faster than `torch.svd` |
| speedup range | near parity on some rows to over `7000x` on best tiny/specialized rows |

The implementation keeps unsupported or numerically risky cases on ATen fallback rather than forcing every shape through the Triton paths.

File: tests/test_svd.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
+import pytest
+import torch
+
+import flag_gems
+
+from . import accuracy_utils as utils
+
+FAST_SHAPES = [(2, 2), (8, 2), (2, 8), (16, 8), (8, 16), (64, 32), (32, 64)]
+FALLBACK_SHAPES = [(5, 3), (3, 5), (2, 4, 4)]
+
```
Comment (bin913) [inline tests/test_svd.py]: Please follow FlagGems to manage test shapes uniformly.

File: tests/test_svd.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    ref_u, ref_s, ref_v = torch.svd(ref_inp, some=True, compute_uv=True)
+
+    with flag_gems.use_gems(include=["svd"]):
+        res_u, res_s, res_v = torch.svd(inp, some=True, compute_uv=True)
+
+    assert res_u.shape == ref_u.shape
+    assert res_s.shape == ref_s.shape
+    assert res_v.shape == ref_v.shape
+    _assert_close(res_s, ref_s, atol=2e-3, rtol=2e-3)
+    _assert_close(_reconstruct(res_u, res_s, res_v), ref_inp, atol=2e-3, rtol=2e-3)
```
Comment (bin913) [inline tests/test_svd.py]: Please use the assert_close function from FlagGems.

File: tests/test_svd.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    _assert_orthonormal(res_u)
+    _assert_orthonormal(res_v)
```
Comment (bin913) [inline tests/test_svd.py:141]: Please use the assert_close function from FlagGems.

File: tests/test_svd.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
```
Comment (bin913) [inline tests/test_svd.py:164]: Please use the assert_close function from FlagGems.

File: tests/test_svd.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        res_u, res_s, res_v = torch.svd(inp, some=True, compute_uv=True)
+
+    assert torch.isfinite(res_u).all()
+    assert torch.isfinite(res_s).all()
+    assert torch.isfinite(res_v).all()
+    assert res_u.shape == ref_u.shape
+    assert res_s.shape == ref_s.shape
+    assert res_v.shape == ref_v.shape
+    _assert_close(res_s, ref_s, atol=2e-3, rtol=2e-3)
+    _assert_close(_reconstruct(res_u, res_s, res_v), ref_inp, atol=2e-3, rtol=2e-3)
```
Comment (bin913) [inline tests/test_svd.py]: Please use the assert_close function from FlagGems.

File: tests/test_svd.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    ref_inp = utils.to_reference(inp, False)
+    ref_u, ref_s, ref_v = torch.svd(ref_inp, some=some, compute_uv=False)
+
+    with flag_gems.use_gems(include=["svd"]):
+        res_u, res_s, res_v = torch.svd(inp, some=some, compute_uv=False)
+
+    assert res_u.shape == ref_u.shape
+    assert res_s.shape == ref_s.shape
+    assert res_v.shape == ref_v.shape
+    _assert_close(res_s, ref_s)
```
Comment (bin913) [inline tests/test_svd.py]: Please use the assert_close function from FlagGems.

File: tests/test_svd.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    for inp in inputs:
+        ref_inp = utils.to_reference(inp, False)
+        ref_u, ref_s, ref_v = torch.svd(ref_inp)
+        with flag_gems.use_gems(include=["svd"]):
+            res_u, res_s, res_v = torch.svd(inp)
+
+        assert res_u.shape == ref_u.shape
+        assert res_s.shape == ref_s.shape
+        assert res_v.shape == ref_v.shape
+        _assert_close(res_s, ref_s, atol=2e-3, rtol=2e-3)
```
Comment (bin913) [inline tests/test_svd.py]: Please use the assert_close function from FlagGems.

File: pytest.ini [hunk 9]
```diff
@@ -5,3 +5,5 @@ filterwarnings =
     ignore::pytest.PytestUnknownMarkWarning
     ignore::DeprecationWarning
     ignore:Warning only once for all operators,  other operators may also be .*:UserWarning
+markers =
+    svd: marks SVD operator tests and benchmarks
```
Comment (bin913) [inline pytest.ini]: Please avoid registering the SVD op here.

File: src/flag_gems/ops/svd.py [hunk 10]
```diff
@@ -0,0 +1,855 @@
...
+
+def _native_linalg_svd(input, some=True, compute_uv=True):
+    if not compute_uv:
+        return _fallback_svd(input, some, compute_uv)
+    u, s, vh = torch.linalg.svd(input.contiguous(), full_matrices=not some)
+    return u, s, vh.mH
+
+
+def _aten_bmm(left, right, out_shape):
+    out = torch.ops.aten.bmm.default.redispatch(_FALLBACK_KEYSET, left, right)
```
Comment (bin913) [inline src/flag_gems/ops/svd.py]: FlagGems is a library implemented with Triton. Please implement the operator using Triton.

File: src/flag_gems/ops/svd.py [hunk 11]
```diff
@@ -0,0 +1,855 @@
...
+
+
+def _is_iluvatar_backend():
+    return device.vendor_name == "iluvatar"
+
+
+def _native_linalg_svd(input, some=True, compute_uv=True):
+    if not compute_uv:
+        return _fallback_svd(input, some, compute_uv)
+    u, s, vh = torch.linalg.svd(input.contiguous(), full_matrices=not some)
```
Comment (bin913) [inline src/flag_gems/ops/svd.py]: FlagGems is a library implemented with Triton. Please implement the operator using Triton.

File: src/flag_gems/ops/svd.py [hunk 12]
```diff
@@ -0,0 +1,855 @@
...
+    torch._C.DispatchKey.CompositeImplicitAutograd
+)
+
+_GRAM_CONDITION_GUARD_MAX_BATCH = 16
+_GRAM_CONDITION_GUARD_MAX_K = 32
+_GRAM_CONDITION_EIGEN_RATIO = 1.0e-8
+
+
+def _fallback_svd(input, some=True, compute_uv=True):
+    return torch.ops.aten.svd.default.redispatch(
```
Comment (bin913) [inline src/flag_gems/ops/svd.py]: FlagGems is a library implemented with Triton. Please implement the operator using Triton.

Comment (bin913) [reply src/flag_gems/ops/svd.py]: > FlagGems is a library implemented with Triton. Please implement the operator using Triton.

There is still torch implementation in _fallback_svd()

**Comment by bin913** [comment]
@lcthej   The torch.ops.aten.svd.default.redispatch is a PyTorch function, and it should not be called within Triton kernels.

**Review by bin913 — APPROVED** [review]
